avoid information loss on .GoError()

tbg opened the pull request:
fixes #2628.

tamird commented on proto/errors.go:
```diff
@@ -67,7 +72,42 @@ func NewError(err error) *Error {
+	if e.Index == nil {
+		return 0, false
+	}
+	return (*Error)(e).Index.Index, true
```
why bother with the cast?

tschottdorf replied:
I think I used `GetIndex()` before. Changed.

tamird commented on storage/store_test.go:
```diff
@@ -1550,3 +1551,19 @@ func TestMaybeRemove(t *testing.T) {
+	pErr := &proto.Error{}
+	pErr.SetResponseGoError(err)
+	if pErr.GetIndex().GetIndex() != ind {
+		log.Warningf("proto.Error lost index information")
```
whats this about?

tschottdorf replied:
this just makes sure that `SetResponseGoError()` will set the `Index` in the `*proto.Error` message. If `err` implemented `ErrorDetail` (doesn't now, but could in the future), the next test using `GoError()` below would get the information from the original error, but it should do the same in either case.

tbg committed:
avoid information loss on .GoError()

fixes #2628.

tbg committed:
(*Error).Set{Response->}GoError

tbg committed:
remove errWithIndex

remove errWithIndex, which was used in `./storage` to wrap a batch
execution error along with the position at which it occurred.
That construct was prone to bugs and the later unwrapping into
a proto.Error meant that you could end up with a proto.Error
which carried more information than its associated ErrorDetail
(for example, `errWithIndex{err: ConditionalPutError, index:1}`
would translate into `proto.Error{Index: 1, Detail: ConditionalPutError}`
but since `ConditionalPutError` is a valid `ErrorDetail`, `GoError()`
would then return exactly that, losing the index.

Now an error whose relative position could be relevant at some future
point needs to implement `proto.IndexedError`, and then everything will
work as expected, regardless of whether the original error is an
`ErrorDetail` or not.

tbg committed:
return `error` from {db,txn}.Run (again)

this is now possible since the index information is
stored on the `ErrorDetail` itself, so exposing the
`GoError()` only is sufficient.

tschottdorf commented:
PTAL. I actually cleaned up some more stuff in the vicinity (see third commit message), which brought some nice simplifications.

tbg committed:
some future proofing

tamird commented:
That these errors carry indexes into the batch in which they are contained is starting to feel pretty dirty to me. What's the long term plan here?

tamird commented on storage/store_test.go:
```diff
@@ -465,8 +465,6 @@ func TestStoreExecuteNoop(t *testing.T) {
 
-// TestStoreErrorWithIndex verifies that an *errWithIndex returned from the
-// replica is correctly translated into a *proto.Error with a corresponding
 // Index. See the companion test TestBatchWithError.
```
this comment is broken now

tschottdorf replied:
Removed the test. `TestBatchErrorWithIndex` tests the same thing closer to
the source.

On Wed, Sep 23, 2015 at 10:53 AM, Tamir Duberstein <notifications@github.com
> wrote:

> In storage/store_test.go
> <https://github.com/cockroachdb/cockroach/pull/2629#discussion_r40212652>:
>
> > @@ -465,8 +465,6 @@ func TestStoreExecuteNoop(t *testing.T) {
> >  	}
> >  }
> >
> > -// TestStoreErrorWithIndex verifies that an *errWithIndex returned from the
> > -// replica is correctly translated into a *proto.Error with a corresponding
> >  // Index. See the companion test TestBatchWithError.
>
> this comment is broken now
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2629/files#r40212652>.
>

tamird commented on proto/errors.go:
```diff
@@ -113,13 +175,22 @@ func (e *Error) SetResponseGoError(err error) {
 		e.Retryable = r.CanRetry()
 	}
+	var isTxnError bool
 	if r, ok := err.(TransactionRestartError); ok {
```
s/ok/isTxnError/

tschottdorf replied:
I need it defined in the outside scope.

tamird replied:
yeah i understand, but you can do that by just:
```go
r, isTxnError := err.(TransactionRestartError)
if isTxnError {
  e.TransactionRestart = r.CanRestartTransaction()
}
```

tschottdorf replied:
going to leave as is to keep the symmetry with the other block.

tamird commented:
LGTM

petermattis commented on proto/errors.proto:
```diff
@@ -192,6 +193,11 @@ enum TransactionRestart {
 }
 
+// ErrPosition describes the position of an error in a Batch.
+message ErrPosition {
```
Is there a value to this message versus just embedding `optional int32 index` in the errors which require it?

tamird replied:
yes, in proto3 primitive types are not nullable. there was a discussion on this in a previous PR

tschottdorf replied:
I've moved up the comment about that from `message Error`.

petermattis commented:
LGTM

bdarnell commented:
LGTM. This change is definitely an improvement, but embedding the position doesn't sit well with me either. Is there something else we could do to provide the same functionality? For example, if ConditionalPut returned the key as well as the value, would that be enough for the SQL layer to construct the right error message for uniqueness violations?

tamird commented:
@bdarnell yes, that would be enough for the SQL layer. However, I think in general we will still want an error -> operation mapping for batches.

tschottdorf commented:
The only location in which they're really "on" the batch is when they're on
the wire, but everywhere else they're separate. In the end, I would like to
completely remove the error field from `*proto.BatchResponse` and just send
a `proto.BatchResponseOrError` over the wire, so that you never see both a
response and an error.

On Wed, Sep 23, 2015 at 10:52 AM, Tamir Duberstein <notifications@github.com
> wrote:

> That these errors carry indexes into the batch in which they are contained
> is starting to feel pretty dirty to me. What's the long term plan here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2629#issuecomment-142626661>
> .
>

tschottdorf commented:
> For example, if ConditionalPut returned the key as well as the value

Yes, I thought about that too but it seemed fairly inflexible. With hindsight, while I think this change makes indexed errors workable, I agree it's not the cream of the crop yet and it might've been better to start small and just add a `Key` to `proto.ConditionalPutError`. Well, damage is done now. I'm sure there's going to be another iteration.

tbg committed:
remove redundant test

plus some comment fixes
